File: 0_training/shopee/math/1.two-sum.java
import java.util.HashMap;
import java.util.HashSet;

/*
 * @lc app=leetcode id=1 lang=java
 *
 * [1] Two Sum
 */

// @lc code=start
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];
            if (map.containsKey(diff)) {
                int res[] = { map.get(diff), i };
                return res;
            }
            map.put(nums[i], i);
        }
        return null;
    }
}
// @lc code=end

File: array_hashing/golang/1.two-sum.go
/*
 * @lc app=leetcode id=1 lang=golang
 *
 * [1] Two Sum
 */

// @lc code=start
func twoSum(nums []int, target int) []int {
    // HashMap Approach
    // Solution by ilham surya 20/07/22

    // Space: O(n)
    hashMap := make(map[int]int)
	
	// Time: O(n)
    for index, num := range nums {
        desired := target - num
        if value, exists := hashMap[desired]; exists {
            return []int{value, index}
        }
        hashMap[num] = index
    }
    return []int{}
}
// @lc code=end

